Cumsum

jameswex opened the pull request:
Implements cumsum operation

dsmilkov commented:
Thanks. Left a few comments - mostly how to do some transposing and rehsaping before you call the webgl shader. This will simplify the work a lot and allow summation across arbitrary axis.

---

Reviewed 5 of 8 files at r1.
Review status: 5 of 8 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/tensor.ts, line 420 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFowSqBOUK95QplyvA:-LCFowSr5ym3jfKpBuO-:b-cab17m) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/tensor.ts#L420)):*
> ```XML
> 
>   /**
>    * Returns a `Tensor` that has expanded rank, by inserting a dimension
> ```

this jsdoc doesn't seem right. Is this taken from expandDim? Did you mean `Compute the cumulative sum of the tensor x along axis.`

---

*[src/tensor.ts, line 423 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFpL3u80ywVxRB3wBy:-LCFpL3u80ywVxRB3wBz:b-m9pqp5) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/tensor.ts#L423)):*
> ```XML
>    * into the tensor's shape. See `expandDims` for details.
>    *
>    * @param axis The dimension index at which to insert shape of 1. Defaults to
> ```

the axis along which to sum.

---

*[src/kernels/webgl/cumsum_gpu.ts, line 49 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFpfz4BBw2Ma_GePGL:-LCFpfz4BBw2Ma_GePGM:b-uv7i6k) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/kernels/webgl/cumsum_gpu.ts#L49)):*
> ```XML
>         ${dtype} adjustableCoords = ${dtype}(${outputCoords});
>         int finalCoord = int(${finalCoord});
>         
> ```

remove whitespace

---

*[src/kernels/webgl/cumsum_gpu.ts, line 68 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFvHZc0RhS-YOeLgEe:-LCFvHZc0RhS-YOeLgEf:b-fty51k) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/kernels/webgl/cumsum_gpu.ts#L68)):*
> ```XML
> }
> 
> function getCoords(rank: number, name: string): string {
> ```

you can simplify this shader a bit if you do the following:
- transpose to make `axis` be the inner-most axis
- reshape to 2D where all other-most shapes are flattened
- shader operates always on a 2d tensor and sums along it's inner-most axis
- after shader runs, undo the reshape, and undo the transpose

---

*[src/ops/array_ops.ts, line 1147 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFtjjrEZQ_HsDRJs6Y:-LCFtjjrEZQ_HsDRJs6Z:b-lm7mup) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops.ts#L1147)):*
> ```XML
>    * x.cumsum().print();
>    * ```
>    *
> ```

add one more code snippet with 2d and sum along axis 0

---

*[src/ops/array_ops.ts, line 1151 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFtqfv-kynVgEjJRmr:-LCFtqfv-kynVgEjJRms:bp7pgcw) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops.ts#L1151)):*
> ```XML
>    * @param axis The axis on which to perform the sum. Optional. Defaults to 0.
>    * @param exclusive Whether to perform exclusive cumulative sum. Optional.
>    *     Defaults to false.
> ```

Expand a bit on what exclusive means.

---

*[src/ops/array_ops.ts, line 1164 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFuMX8-dgqM4mLOJAQ:-LCFuMX8-dgqM4mLOJAR:b-66km60) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops.ts#L1164)):*
> ```XML
> 
>     return ENV.engine.runKernel(
>                backend => backend.cumsum(x, axis, exclusive, reverse), {x},
> ```

this won't work for axis that are not inner most unless you transpose to make the axis be inner-most, do the computation and then transpose back again. See https://github.com/tensorflow/tfjs-core/blob/master/src/ops/reduction_ops.ts#L118 for how to transpose.

---

*[src/ops/array_ops_test.ts, line 2295 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFvpbFAfPL-6VjuWpN:-LCFvpbFAfPL-6VjuWpO:b-i31npb) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops_test.ts#L2295)):*
> ```XML
>   });
> 
>   it('2D reverse exclusive', () => {
> ```

add test for 2d with axis=0

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-:-LCFvtXy127m9zROnts6:b-xw0eng)*
<!-- Sent from Reviewable.io -->

jameswex commented:
added the transpose so that it works on any axis.

---

Review status: 5 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[src/tensor.ts, line 420 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFowSqBOUK95QplyvA:-LCTr-WRAoo4DLsYOfFt:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/tensor.ts#L420)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

this jsdoc doesn't seem right. Is this taken from expandDim? Did you mean `Compute the cumulative sum of the tensor x along axis.`
</blockquote></details>

Done.

---

*[src/tensor.ts, line 423 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFpL3u80ywVxRB3wBy:-LCTr8SvEH-v322R27Bd:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/tensor.ts#L423)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

the axis along which to sum.
</blockquote></details>

Done.

---

*[src/kernels/webgl/cumsum_gpu.ts, line 49 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFpfz4BBw2Ma_GePGL:-LCTtZdJARwEH0ZAVe41:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/kernels/webgl/cumsum_gpu.ts#L49)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

remove whitespace
</blockquote></details>

Done.

---

*[src/kernels/webgl/cumsum_gpu.ts, line 68 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFvHZc0RhS-YOeLgEe:-LCUT3lyBUCV9SuUZH1F:bbllx0f) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/kernels/webgl/cumsum_gpu.ts#L68)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

you can simplify this shader a bit if you do the following:
- transpose to make `axis` be the inner-most axis
- reshape to 2D where all other-most shapes are flattened
- shader operates always on a 2d tensor and sums along it's inner-most axis
- after shader runs, undo the reshape, and undo the transpose

</blockquote></details>

added transpose/undo transpose in array_ops so it applies for both cpu and gpu.

---

*[src/ops/array_ops.ts, line 1147 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFtjjrEZQ_HsDRJs6Y:-LCTs75I3QyYZayQQEEX:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops.ts#L1147)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

add one more code snippet with 2d and sum along axis 0
</blockquote></details>

Done.

---

*[src/ops/array_ops.ts, line 1151 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFtqfv-kynVgEjJRmr:-LCTrAKsBj2lbbivMjdt:b-7rl0kd) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops.ts#L1151)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Expand a bit on what exclusive means.
</blockquote></details>

done

---

*[src/ops/array_ops.ts, line 1164 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFuMX8-dgqM4mLOJAQ:-LCUTBcf7wtGRhty5jCI:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops.ts#L1164)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

this won't work for axis that are not inner most unless you transpose to make the axis be inner-most, do the computation and then transpose back again. See https://github.com/tensorflow/tfjs-core/blob/master/src/ops/reduction_ops.ts#L118 for how to transpose.
</blockquote></details>

Done.

---

*[src/ops/array_ops_test.ts, line 2295 at r1](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCFvpbFAfPL-6VjuWpN:-LCTtLLuCvNYi2HyG64w:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/efcf58143afcb3b1636bea5085872f5605830358/src/ops/array_ops_test.ts#L2295)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

add test for 2d with axis=0
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-:-LCUYo785CizYNjKIOA_:b-57g6p1)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r3.
Review status: 2 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/kernels/webgl/cumsum_gpu.ts, line 3 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZEaK852rxh-UTZYYJ:-LCZEaK852rxh-UTZYYK:bb6ht2i) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/kernels/webgl/cumsum_gpu.ts#L3)):*
> ```XML
> /**
>  * @license
>  * Copyright 2017 Google Inc. All Rights Reserved.
> ```

2018

---

*[src/ops/array_ops.ts, line 1188 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZ7wU13yaErw3JaPJQ:-LCZ7wU24VhPTbNH2MSH:brie8i) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/ops/array_ops.ts#L1188)):*
> ```XML
>     };
>     let value =  ENV.engine.runKernel(
>       backend => backend.cumsum(permutedX, axis, exclusive, reverse),
> ```

indent 2 more

---

*[src/ops/array_ops_test.ts, line 2349 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZ8JdF12EpcxS0t4PL:-LCZ8JdGEWpUZJOsct4_:b8ore5c) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/ops/array_ops_test.ts#L2349)):*
> ```XML
> 
>   it('1D reverse', () => {
>     const res = tf.tensor1d([1, 2, 3, 4]).cumsum(0, false, true);
> ```

pull these false / true to variables for readability

---

*[src/ops/array_ops_test.ts, line 2394 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZ8EX-6NksUiRE5EtF:-LCZ8EX-6NksUiRE5EtG:bha3ecz) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/ops/array_ops_test.ts#L2394)):*
> ```XML
> 
>   it('3D standard', () => {
>     const res =tf.tensor3d([[[0, 1], [2, 3]], [[4, 5], [6, 7]]]).cumsum(2);
> ```

spacing here (btw if you use vscode this will auto format for you)

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-:-LCZEjIA93nGQee_tEVA:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

jameswex commented:
Review status: 2 of 8 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[src/kernels/webgl/cumsum_gpu.ts, line 3 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZEaK852rxh-UTZYYJ:-LCZH3zs2tJGQYF1Ezig:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/kernels/webgl/cumsum_gpu.ts#L3)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

2018
</blockquote></details>

Done.

---

*[src/ops/array_ops.ts, line 1188 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZ7wU13yaErw3JaPJQ:-LCZHHzL-_eQHfDoDj60:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/ops/array_ops.ts#L1188)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

indent 2 more
</blockquote></details>

Done.

---

*[src/ops/array_ops_test.ts, line 2349 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZ8JdF12EpcxS0t4PL:-LCZHaNNBcPh96KrStyn:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/ops/array_ops_test.ts#L2349)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

pull these false / true to variables for readability
</blockquote></details>

Done.

---

*[src/ops/array_ops_test.ts, line 2394 at r4](https://reviewable.io/reviews/tensorflow/tfjs-core/1032#-LCZ8EX-6NksUiRE5EtF:-LCZHcLQ5AunB1e6cOb6:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-core/blob/dfdb3bfc7a3c767c1a3bf4b9a42332ff40b7d9bd/src/ops/array_ops_test.ts#L2394)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

spacing here (btw if you use vscode this will auto format for you)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-core/1032)*
<!-- Sent from Reviewable.io -->
